pennylane/ops/functions/bind_new_parameters.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the qml.bind_new_parameters function.
+"""
+
+from typing import Sequence
+import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.typing import TensorLike
+from pennylane.operation import Operator, Tensor
+
+from ..op_math import CompositeOp, SymbolicOp, ScalarSymbolicOp, Adjoint, Pow, SProd
+
+
+@singledispatch
+def bind_new_parameters(op: Operator, params: Sequence[TensorLike]) -> Operator:
+    """Create a new operator with updated parameters
+
+    This function takes an :class:`~.Operator` and new parameters as input and
+    returns a new operator of the same type with the new parameters. This function
+    does not mutate the original operator.
+
+    Args:
+        op (.Operator): Operator to update
+        params (Sequence[TensorLike]): New parameters to create operator with
+
+    Returns:
+        .Operator: New operator with updated parameters
+
+    Raises:
+        ValueError: If the shape of the old and new operator parameters don't match
+    """
+
+    return op.__class__(*params, wires=op.wires, **copy.deepcopy(op.hyperparameters))
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_composite_op(op: CompositeOp, params: Sequence[TensorLike]):
+    new_operands = []
+
+    for operand in op.operands:
+        op_num_params = operand.num_params
+        sub_params = params[: op_num_params]
+        params = params[op_num_params :]
+        new_operands.append(bind_new_parameters(operand, sub_params))
+
+    return op.__class__(*new_operands)
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_symbolic_op(op: SymbolicOp, params: Sequence[TensorLike]):
+
+    new_base = bind_new_parameters(op.base, params)
+    new_hyperparameters = copy.deepcopy(op.hyperparameters)
+    _ = new_hyperparameters.pop("base")
+
+    if isinstance(op, Adjoint):
+        # Need this branch because using the other class signature results in a call to
+        # `Adjoint.__new__` which doesn't raise an error but does return an unusable
+        # object
+        return Adjoint(new_base)
```

Comment:
Should we just register a specialized method for `Adjoint`?

Reply:
I thought this comment meant that a special Adjoint method causes problems, but if I misunderstood then yes please register it